File: examples/ciflypy-examples/ciflypy_examples/optimal_adjustment.py
import ciflypy as cf
import ciflypy_examples.utils as utils

ruletables = utils.get_ruletable_path()
not_amenable_table = cf.Ruletable(ruletables / "not_amenable_cpdag.txt")
possible_ancestors_table = cf.Ruletable(ruletables / "possible_ancestors_cpdag.txt")
possible_descendants_table = cf.Ruletable(ruletables / "possible_descendants_cpdag.txt")
parents_table = cf.Ruletable(ruletables / "parents_cpdag.txt")


def optimal_adjustment(cpdag, X, Y):
    descendants = cf.reach(cpdag, {"X": X}, possible_descendants_table)
    if set(Y).difference(descendants):
        return None

    not_amenable = cf.reach(cpdag, {"X": X}, not_amenable_table)
    ancestors = cf.reach(cpdag, {"X": Y, "W": X}, possible_ancestors_table)
    cn = set(ancestors).intersection(descendants)
    forbidden = cf.reach(cpdag, {"X": cn}, possible_descendants_table)
    if set(forbidden).intersection(X) or set(Y).intersection(not_amenable):
        return None

    pre_opt = cf.reach(cpdag, {"X": cn}, parents_table)
    return set(pre_opt).difference(set(forbidden).union(X))

File: examples/ciflyr-examples/R/optimalAdjustment.R
library("ciflyr") 
library("here")

source(here("R", "utils.R"))

ruletables <- getRuletablePath()
notAmenable <- parseRuletable(file.path(ruletables, "not_amenable_cpdag.txt"))
possibleAncestors <- parseRuletable(file.path(ruletables, "possible_ancestors_cpdag.txt"))
possibleDescendants <- parseRuletable(file.path(ruletables, "possible_descendants_cpdag.txt"))
parentsTable <- parseRuletable(file.path(ruletables, "parents_cpdag.txt"))

optimalAdjustment <- function(cpdag, X, Y){
	des <- reach(cpdag, list("X" = X), possibleDescendants)
	if (length(intersect(Y, des)) < length(Y)) {
		return (NULL)
	}
	
	nam <- reach(cpdag, list("X" = X), notAmenable)
	anc <- reach(cpdag, list("X" = Y, "W" = X), possibleAncestors)
	cn <- intersect(anc, des)
	forb <- reach(cpdag, list("X" = cn), possibleDescendants)
	if (length(intersect(forb, X)) != 0 | length(intersect(nam, Y)) != 0) {
		return (NULL)
	}
	
	pre_opt <- reach(cpdag, list("X" = cn), parentsTable)
	
	return (setdiff(pre_opt, union(forb, X)))
}
